examples/word-count/dev/user.clj

```diff
@@ -0,0 +1,118 @@
+     :replication-factor 1
+     :key-serde (js/edn-serde)
+     :value-serde (js/edn-serde)})
+
+  Associative
+  (assoc [this key val]
+    this))
+
+(defn new-fake-topic-metadata []
+  (FakeTopicMetadata.))
+
+(def topic-metadata
+  (new-fake-topic-metadata))
+
+
+(defn topology->topic-metadata
+  [topology streams-config]
+  (->> (jsd/describe-topology (.build (j/streams-builder* topology)) streams-config)
+       (map :nodes)
+       (reduce concat)
+       (filter #(= :topic (:type %)))
+       (remove (fn [x] (re-matches #".*STATE-STORE.*" (:name x))))
+       (map :name)
+       (reduce (fn [acc x] (assoc acc (keyword x) (get (new-fake-topic-metadata) x))) {})))
+
+(defn re-delete-topics
+  "Takes an instance of java.util.regex.Pattern and deletes any Kafka
+  topics that match."
+  [config re]
+  (with-open [client (ja/->AdminClient config)]
+    (let [topics-to-delete (->> (ja/list-topics client)
+                                (filter #(re-find re (:topic-name %))))]
+      (ja/delete-topics! client topics-to-delete))))
+
+(defn destroy-state-stores
+  "Takes an application config and deletes local files associated with
+  internal state."
+  [streams-config]
+  (sh "rm" "-rf" (str "/tmp/kafka-streams/" (:application-id streams-config))))
+
+
+(defmethod ig/init-key :streams-config [_ streams-config]
+  (let [bootstrap-servers (or (System/getenv "BOOTSTRAP_SERVERS")
+                              (:bootstrap-servers streams-config))]
+    (assoc streams-config :bootstrap-servers bootstrap-servers)))
+
+(defmethod ig/init-key :topology [_ {:keys [topology-builder-fn]}]
+  (let [topology-builder (deref (find-var (symbol topology-builder-fn)))
+        builder (j/streams-builder)]
+    ((topology-builder (new-fake-topic-metadata)) builder)))
+
+(defmethod ig/init-key :topics [_ {:keys [streams-config topology] :as opts}]
+  (let [config (word-count/propertize (select-keys streams-config [:bootstrap-servers]))
+        topic-metadata (topology->topic-metadata topology streams-config)]
+    (with-open [client (ja/->AdminClient config)]
+      (ja/create-topics! client (vals topic-metadata)))
+    (assoc opts :topic-metadata topic-metadata)))
+
+(defmethod ig/init-key :app [_ {:keys [streams-config topology] :as opts}]
+  (let [streams-app (j/kafka-streams topology (word-count/propertize streams-config))]
+    (j/start streams-app)
+    (assoc opts :streams-app streams-app)))
+
+
+(defmethod ig/halt-key! :topics [_ {:keys [streams-config topic-metadata]}]
+  (let [config (word-count/propertize (select-keys streams-config [:bootstrap-servers]))]
+    (re-delete-topics config (re-pattern (str "(" (->> topic-metadata
+                                                       keys
+                                                       (map name)
+                                                       (str/join "|"))
+                                              ")")))))
+
+(defmethod ig/halt-key! :app [_ {:keys [streams-config streams-app]}]
```

Comment:
I started with that, and then separated out the cleanup or resources because sometimes you just want to stop the app but not delete everything so you can inspect the topics and state stores. https://github.com/FundingCircle/jackdaw-repl/commit/0ba5f6f70ab2a08f8f4d11819c5540d3e093574c

Reply:
Seems like you'd want a full reset to be able to start clean. Could we halt just the app with a different config structure?